core/Tracker/Failures.php

```diff
@@ -0,0 +1,178 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\Tracker;
+
+use Piwik\Common;
+use Piwik\Date;
+use Piwik\Exception\UnexpectedWebsiteFoundException;
+use Piwik\Piwik;
+use Piwik\Site;
+use Piwik\Db as PiwikDb;
+
+class Failures
+{
+    const CLEANUP_OLD_FAILURES_DAYS = 2;
+    const FAILURE_ID_INVALID_SITE = 1;
+    const FAILURE_ID_NOT_AUTHENTICATED = 2;
+
+    private $table = 'tracking_failure';
+    private $tablePrefixed;
+    private $now;
+
+    public function __construct()
+    {
+        $this->tablePrefixed = Common::prefixTable($this->table);
+    }
+
+    public function setNow(Date $now)
+    {
+        $this->now = $now;
+    }
+
+    private function getNow()
+    {
+        if (isset($this->now)) {
+            return $this->now;
+        }
+        return Date::now();
+    }
+
+    public function logFailure($idFailure, Request $request)
+    {
+        $visitExcluded = new VisitExcluded($request);
+        if ($visitExcluded->isExcluded()) {
+            return;
+        }
```

Comment:
`VisitExcluded::isExcluded()` is called during tracking correct? If it were possible to set the result in `VisitProperties`, we could use that value here instead of checking if the request should be excluded again. Might be a bit faster?